File: DurableConsumptionModel/negm.py
import numpy as np
from numba import njit, prange

# consav
from consav import linear_interp # for linear interpolation
from consav import upperenvelope

# local modules
import utility

negm_upperenvelope = upperenvelope.create(utility.func,use_inv_w=True)

@njit(parallel=True)
def solve_keep(t,sol,par):
    """solve the bellman equation using the endogenous grid method"""

    # unpack
    inv_v = sol.inv_v_keep[t]
    inv_marg_u = sol.inv_marg_u_keep[t]
    c = sol.c_keep[t]
    q_c = sol.q_c[t]
    q_m = sol.q_m[t]

    for i_p in prange(par.Np):
        
        # temporary container
        v_ast_vec = np.zeros(par.Nm)

        for i_n in range(par.Nn):
            
            # use euler equation
            n = par.grid_n[i_n]
            for i_a in range(par.Na):
                q_c[i_p,i_n,i_a] = utility.inv_marg_func(sol.q[t,i_p,i_n,i_a],n,par)
                q_m[i_p,i_n,i_a] = par.grid_a[i_a] + q_c[i_p,i_n,i_a]
        
            # upperenvelope
            negm_upperenvelope(par.grid_a,q_m[i_p,i_n],q_c[i_p,i_n],sol.inv_w[t,i_p,i_n],
               par.grid_m,c[i_p,i_n],v_ast_vec,n,par)        

            # negative inverse
            for i_m in range(par.Nm):
                inv_v[i_p,i_n,i_m] = -1/v_ast_vec[i_m]
                if par.do_marg_u:
                    inv_marg_u[i_p,i_n,i_m] = 1/utility.marg_func(c[i_p,i_n,i_m],n,par)

File: 02. DurableConsumptionModel/cppfuncs/negm.cpp
#ifndef MAIN
#include "header.cpp"
#endif

EXPORT void compute_wq_negm(par_struct *par, sol_struct *sol)
{
    post_decision::compute_wq(par->t,sol,par,true);
}

EXPORT void solve_negm_keep(par_struct *par, sol_struct *sol)
{

    // unpack
    int t = par->t;
    
    int index_keep_t = index::d4(t,0,0,0,par->T,par->Np,par->Nn,par->Nm);
    auto inv_v = &sol->inv_v_keep[index_keep_t];
    auto inv_marg_u = &sol->inv_marg_u_keep[index_keep_t];
    auto c = &sol->c_keep[index_keep_t];

    int index_post_t = index::d4(t,0,0,0,par->T-1,par->Np,par->Nn,par->Na);
    auto inv_w = &sol->inv_w[index_post_t];
    auto q = &sol->q[index_post_t];
    auto q_c = &sol->q_c[index_post_t];
    auto q_m = &sol->q_m[index_post_t];

    #pragma omp parallel num_threads(par->cppthreads)
    {
        
        // temporary containers
        auto v_ast_vec = new double[par->Nm];

    #pragma omp for
    for(int i_p = 0; i_p < par->Np; i_p++){

        for(int i_n = 0; i_n < par->Nn; i_n++){
            
            // use euler equation
            double n = par->grid_n[i_n];
            for(int i_a = 0; i_a < par->Na; i_a++){
                int index_post = index::d3(i_p,i_n,i_a,par->Np,par->Nn,par->Na);
                q_c[index_post] = utility::inv_marg_func(q[index_post],n,par);
                q_m[index_post] = par->grid_a[i_a] + q_c[index_post];
            }
        
            // upperenvelope
            int index_keep = index::d3(i_p,i_n,0,par->Np,par->Nn,par->Nm);
            int index_post = index::d3(i_p,i_n,0,par->Np,par->Nn,par->Na);
            upperenvelope(par->grid_a,par->Na,&q_m[index_post],&q_c[index_post],&inv_w[index_post],true,
                          par->grid_m,par->Nm,
                          &c[index_keep],v_ast_vec,n,par);  

            // negative inverse
            for(int i_m = 0; i_m < par->Nm; i_m++){
                inv_v[index_keep+i_m] = -1.0/v_ast_vec[i_m];
                if(par->do_marg_u){
                    inv_marg_u[index_keep+i_m] = 1.0/utility::marg_func(c[index_keep+i_m],n,par);
                }
            }
        
        } // n
    
    } // p

        delete[] v_ast_vec;

    } // parallel

} // solve_keep